lib/omniauth/strategies/indieauth.rb

```diff
@@ -2,76 +2,131 @@
 
 require 'omniauth'
 require 'faraday'
+require 'faraday_middleware'
+require 'indieauth_discovery/profile'
 require 'cgi'
 
+require 'omniauth/indieauth/errors'
+
 module OmniAuth
   module Strategies
     # IndieAuth strategy for OmniAuth.
     class IndieAuth
       include OmniAuth::Strategy
 
-      option :name, 'indieauth'
-      option :server, 'https://indieauth.com'
+      option :name, :indieauth
       option :client_id
 
-      attr_accessor :me
-
-      def redirect_uri
-        full_host + script_name + callback_path
-      end
+      attr_reader :profile
 
       def request_phase
-        puts redirect_uri
-        redirect "#{options.server}/sign-in?redirect_uri=#{URI.encode_www_form_component(redirect_uri)}&client_id=#{URI.encode_www_form_component(options.client_id)}"
+        valid_request_phase_params? ? authorize : profile_url_form
       end
 
       def callback_phase
-        puts request.params.inspect
+        @profile = IndieAuthDiscovery::Profile.discover(profile_url)
+        super if authenticate
+      end
 
-        conn = Faraday.new(url: "#{options.server}/auth") do |faraday|
-          faraday.request :url_encoded # form-encode POST params
-        end
-        response = Faraday.post "#{options.server}/auth", {
-          code: request.params['code'],
-          client_id: options.client_id,
-          redirect_uri: redirect_uri
-        }
-        puts response.body
+      uid { authentication_response.body['me'] }
 
-        data = CGI.parse response.body
+      info { authentication_response.body['profile'] || {} }
 
-        if !data['me'].empty?
-          @me = data['me'][0]
-        else
-          fail!(data['error'][0].to_sym, CallbackError.new(data['error'][0].to_sym, data['error_description'][0]))
-        end
+      private
+
+      attr_reader :authentication_response
 
-        super
+      def client_id
+        options.client_id || "#{request.base_url}/"
       end
 
-      uid do
-        @me
+      def valid_request_phase_params?
+        String(request.params['me']).strip != ''
```

Comment:
If ActiveSupport is anywhere in the dependency chain, you could maybe use `blank?` or `present?` here.

Reply:
In this case I wanted to make sure I didn't require ActiveSupport, as people might use this with something simple in Sinatra for example.